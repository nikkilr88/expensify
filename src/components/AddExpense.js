import React, { Component } from 'react'

class AddExpense extends Component {
  render() {
    return <p>Add</p>
  }
}

export default AddExpense
