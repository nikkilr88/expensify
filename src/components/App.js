import React, { Component } from 'react'

class App extends Component {
  render() {
    return (
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo, error?
        Ducimus quaerat saepe quisquam eveniet eius consequuntur. Voluptatibus
        delectus sint voluptas quibusdam in quae natus sed! Sunt facilis
        repudiandae quaerat?
      </p>
    )
  }
}

export default App
