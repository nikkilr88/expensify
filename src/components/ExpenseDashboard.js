import { Link } from 'react-router-dom'

const ExpenseDashboard = () => <p>this is the dashboard</p>

export default ExpenseDashboard
