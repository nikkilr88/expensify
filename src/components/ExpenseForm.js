import React, { Component } from 'react'

class ExpenseForm extends Component {
  state = {
    description: '',
    amount: '',
    note: ''
  }

  onChange = e => {
    e.persist()
    this.setState(() => ({
      [e.target.name]: e.target.value
    }))
  }

  onAmountChange = e => {
    const amount = e.target.value
    if (amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }))
    }
  }

  render() {
    return (
      <div>
        <form>
          <input
            name='description'
            type='text'
            placeholder='Description'
            value={this.state.description}
            onChange={this.onChange}
            autoFocus
          />
          <input
            name='amount'
            type='text'
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder='Amount'
          />
          <textarea
            name='note'
            value={this.state.note}
            onChange={this.onChange}
            placeholder='Add a note for you expense (optional)'
          />

          <button>Save Expense</button>
        </form>
      </div>
    )
  }
}

export default ExpenseForm
