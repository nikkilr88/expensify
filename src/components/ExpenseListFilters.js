import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters'

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }))
  }

  onTextChange = e => {
    this.props.setTextFilter(e.target.value)
  }

  onSelectChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else {
      this.props.sortByAmount()
    }
  }

  render() {
    const { filters } = this.props

    return (
      <div>
        <input type='text' value={filters.text} onChange={this.onTextChange} />
        <select value={filters.sortBy} onChange={this.onSelectChange}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate}
          endDate={filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = state => ({
  filters: state.filters
})

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters)
