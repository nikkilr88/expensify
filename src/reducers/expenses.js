const expensesState = []

const expensesReducer = (state = expensesState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
    case 'REMOVE_EXPENSE':
      return state.filter(exp => exp.id !== action.id)
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        }
      })
    default:
      return state
  }
}

export default expensesReducer
