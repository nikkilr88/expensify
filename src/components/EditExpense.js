import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

const EditExpense = props => {
  return (
    <div>
      <p>Edit expense {props.match.params.id}</p>

      <ExpenseForm
        expense={props.expense}
        onSubmit={expense => {
          props.dispatch(editExpense(props.expense.id, expense))
          props.history.push('/dashboard')
        }}
      />

      <button
        onClick={() => {
          props.dispatch(removeExpense({ id: props.expense.id }))
          props.history.push('/dashboard')
        }}
      >
        Remove
      </button>
    </div>
  )
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

export default connect(mapStateToProps)(EditExpense)
