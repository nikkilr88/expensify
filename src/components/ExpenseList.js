import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import { getVisibleExpenses } from '../selectors/expenses'

export const ExpenseList = ({ expenses }) => {
  return (
    <div>
      {expenses.length === 0
        ? ''
        : expenses.map((expense, i) => (
            <ExpenseListItem key={expense.id} {...expense} />
          ))}
    </div>
  )
}

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList)
