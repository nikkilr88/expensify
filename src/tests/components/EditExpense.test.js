import React from 'react'
import { shallow } from 'enzyme'
import { EditExpense } from '../../components/EditExpense'
import expenses from '../fixtures/expenses'

let editExpense, removeExpense, history, wrapper, expense, match

beforeEach(() => {
  editExpense = jest.fn()
  removeExpense = jest.fn()

  history = {
    push: jest.fn()
  }

  expense = expenses[0]
  match = { params: { id: expense.id } }

  wrapper = shallow(
    <EditExpense
      match={match}
      expense={expense}
      history={history}
      editExpense={editExpense}
      removeExpense={removeExpense}
    />
  )
})

test('should render EditExpense correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense)

  expect(history.push).toHaveBeenLastCalledWith('/dashboard')
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense)
})

test('should handle removeExpense correctly', () => {
  wrapper.find('button').simulate('click')

  expect(history.push).toHaveBeenLastCalledWith('/dashboard')
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expense.id })
})
