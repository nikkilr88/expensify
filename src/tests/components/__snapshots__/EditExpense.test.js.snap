// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpense correctly 1`] = `
<div>
  <p>
    Edit expense 
    1
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 150000,
        "createdAt": 0,
        "description": "rent",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
